### ISSUE TITLE
Margo Preview Release 1 Scope: Workload Management Interface

### ISSUE THREAD

**Issue by ajcraig** [margo/specification#45]
# Purpose
The goal of this Issue is to establish the Margo Preview Release 1 scope. The scope snippet below covers the Workload Management Interface. The goal of this issue is to gain approval on the proposed Preview Release 1 scope. Additional issues will be released for approval regarding the full scope. 

** Updated 11/13/2024 to include details about interoperability scenarios for this section of the specification. 

# Workload Management Interface Scope
1. The ability for the workload orchestration service to provide deployment specification(s) that follow the Margo standard. [Link](https://specification.margo.org/margo-api-reference/workload-api/desired-state-api/desired-state/)
    - This features enabled via the deployment specification include:
        - Deploy one to many workloads
        - Delete one to many existing workloads
        - Update workload(s) to a newer version of source code(container)
        - Reconfigure currently running workload(s) via parameters
2. The ability to build a Workload Orchestration Management service for orchestrating workloads to Margo compliant devices. [Link](https://specification.margo.org/margo-api-reference/margo-api-specification/)
    - Host a API server endpoint that:
        - Receive the workload deployment status updates
        - Receive device capabilities
    - Host a Gitops enabled repository for providing desired state to the edge devices:
        - Deployment specification storage
        - Deployment specification retrieval

3. Device Workload Management Interface 
- The ability to build a Rest API Client that enables the features below:
    - Send the workload deployment status updates as defined in the current specification
    - Send device capability information as defined in the current specification
        - Upon device reconfiguration, the interface should update the workload orchestrator of the new or changed capabilities.
- The ability to use a Git client or library to enable enable the following features:  
    - Retrieve deployment specification associated with the device
    - Configure a polling schedule

# What does interoperability mean to Margo/what personas are impacted?
- Device Vendor
    - Margo compliant devices are compatible with all Margo compliant WOS vendors. 
    - Ensures device vendors are not responsible with supporting various edge "agents".
- Application Vendor
    - Applications built via Margo standard shall be supported on Margo compliant devices.
    - Freedom to only support 1 or both Deployment manifests(compose/helm)
- Workload Orch Software   
    - Margo compliant WOS vendors are compatible with all Margo compliant devices. 
- End User
    - End user has freedom to choose their choice of device and WOS solution based on features and functionality. This is with the knowledge, if they purchase Margo compliant devices they retain the freedom to switch. 
    - Avoids vendor lockin. If End user is not happy with either Device or WOS, they are able to change vendors without having to replace all components in the architecture. 

> **\*** Secure communication is not a priority for this release, although some details might be provided. Secure connectivity will be required in subsequent releases.

## Decisions:
- Do we go with the initial implementation of OpenGitops, which is currently detailed in the specification, or work through various technical options before release.
- The following envisioned features are not included within the preview release 1:
    - Auditability and tracking history of deployment specification IDs
    - Rollback functionality
    - Disaster recovery requirements
    - Maintenance of the workload orchestration service components
    - Enabling container runtime configuration changes. 

# Review/Approval Options:
@margo/approvers Please respond below and consolidate your company's response to a single reply. If additional information is needed to provide your response, please reach out.

Response Options:

- [x] Approve Proposal
- [ ] Reject Proposal - Provide reasoning for rejection
- [ ] Abstain from Vote - No strong opinion either way

**Comment by merrill-harriman-se** [comment]
Does this statement imply that GitOps will be required by the device to get access to the deployment spec?  
"Host a Gitops enabled repository for providing desired state to the edge devices:
      Deployment specification storage
      Deployment specification retrieval"

I am not yet comfortable with requiring GitOps at the end device - not until we have fully analyzed and justified this over a simple REST interface.  I suspect REST is simpler, less resource requirements, no need to open yet another network path, smaller attack surface, less maintenance, and can very likely provide all the needed functionality.

**Comment by effndc** [comment]
On behalf of Intel I am submitting an approval.

**Comment by phil-abb** [comment]
> I am not yet comfortable with requiring GitOps at the end device - not until we have fully analyzed and justified this over a simple REST interface.

@merrill-harriman-se Would you consider this a blocker for any preview release, or would you be ok if there was a statement somewhere indicating we are still looking into how the desired state is handled and the use of GitOps may change?

**Comment by abbottjd** [comment]
From ABB, I approve this proposal for the PR1 scope. For the preview release, I don't see a problem with basing on GitOps but acknowledge that it makes sense to continue exploring alternative methods for storing/retrieving desired state before progressing to a formal specification release.

**Comment by srberard** [comment]
I have similar concerns regarding GitOps at the end device.  While GitOps is a powerful deployment technique/paradigm, I feel that requiring it as part of Margo would add complexity, impede implementation, and be overly prescriptive.

A core strength of systems like Kubernetes is their flexibility.  K8s does not, itself, prescribe GitOps but instead provides APIs that allow GitOps workflows via tools such as ArgoCD, FluxCD, etc.  For those that desire to use GitOps, they can select the appropriate tool that fits their needs.  I feel the Margo should strive to do something similar.  That is, provide mechanisms to enable GitOps but not require it directly.

**Comment by phil-abb** [comment]
>I have similar concerns regarding GitOps at the end device.

@srberard Same question to you.  Would you consider this a blocker for any preview release, or would you be ok if there was a statement somewhere indicating we are still looking into how the desired state is handled and the use of GitOps may change?

**Comment by srberard** [comment]
I'd rather this not be part of the Preview.  While I don't want to block progress, I think including it in the Preview gives the impression that this will be supported.

**Comment by securidee** [comment]
There is also no mention of a GitOps in the Release 1 scope for device requirements.  That seems like a conflict between the workload management interface scope and the device scope.

**Comment by phil-abb** [comment]
>There is also no mention of a GitOps in the Release 1 scope for device requirements.

@securidee I see this part of [3. The ability to host the Margo management interface service(s) that enable workload deployment and maintenance.](https://github.com/margo/specification/issues/44) because for this to be possible, the implementation the device vendor chooses to use would need to be able to work with a Git repository. 

Do you see it differently? What do you feel should be added?

**Comment by securidee** [comment]
> > There is also no mention of a GitOps in the Release 1 scope for device requirements.
> 
> @securidee I see this part of [3. The ability to host the Margo management interface service(s) that enable workload deployment and maintenance.](https://github.com/margo/specification/issues/44) because for this to be possible, the implementation the device vendor chooses to use would need to be able to work with a Git repository.
> 
> Do you see it differently? What do you feel should be added?

You are correct.  Sorry, I missed the nested requirements in [3. The ability to host the Margo management interface service(s) that enable workload deployment and maintenance.](https://github.com/margo/specification/issues/44) I mistakenly thought a device vendor could just read that portion of the spec, but the Device Workload Management Interface has more specific requirements that apply to the device.

**Comment by stormc** [comment]
I do see the need to go forward with a reasonable snapshot/prototype technical incarnation as Preview Release 1, knowing and clearly communicating that all technical decisions/specifications and hence this realization is most likely subject to change in a later (Preview) Release due to the gist of this preview release (from a 10k ft perspective): Give Application developers something to play with in order to provide feedback for the Margo "look+feel". 

Following this argumentation, parts of the realization (e.g. GitOps vs. REST) are "just" implementation details that will most likely look/be different in the releases to come. Nonetheless, I think we need to come to a consensus on whether or not we can go forward with GitOps in this Preview Release 1, considering the objections in this thread which I second.

Having said that, the vote is: Abstain from Vote.

In particular for the GitOps vs. REST discussion at hand, I'd like to bring in @Silvanoc into the discussion who is very proficient (not only) in this topic. And I'd like to drop some reading material he shared with me:
https://sestegra.medium.com/gitops-with-flux-leveraging-oci-registry-as-a-state-store-ffe28dd64836
https://fluxcd.io/flux/cheatsheets/oci-artifacts/

**Comment by hsinghcg** [comment]
Capgemini View: Abstain from vote - no strong opinion either way.
Not opposed to using gitops in the Preview Release 1 though. Would like us to continue with our analysis on whether or not to _impose_ gitops as the standard (thereby imposing some requirements for the device and the WOS).

This is also related to the topic of what 'range of devices' we'd intend to support - and what's the extreme in terms of the smallest device running linux with kubernetes or docker - and possibly to use that as a benchmark when we decide to impose technology (+ related default tooling or sw stack) choices.

**Comment by merrill-harriman-se** [comment]
I do not agree with a standards-based specification having a requirement for a protocol and API that are not themselves derived from a standard.  GitOps is not a standard - it is a tool and even OpenGitOps is, at best, a DeFacto standard.  This implies that there is no way to specify an API version of the "GitOps Standard" as it does not exist.  It could change underneath our developments later and we cannot do anything to indicate what we support or not.  This, on its own merit, prevents the requirement of GitOps in our standard.  How will we have any form of conformance testing?  Conformant to what?

**Comment by srberard** [comment]
I concur with @merrill-harriman-se, while we speak of GitOps, it is not a formal standard, it is a methodology and set of principles.  

Further, as I stated previously, I am concerned that this will add complexity to Margo and impede progress.  GitOps has been a point of discussion and debate for many weeks now and I believe is distracting us from moving forward.  As the above proposal already requires a REST API interface, I would suggest we limit this proposal to that only and drop GitOps for now.  We can revisit this later.

As such, I reject the proposal as written with the above reasoning.

**Comment by pauldbrooks** [comment]
Rockwell Automation approves with comment
Using Gitops for initial preview release is OK, but does not constrain other options being added either during preview release development, or in final specifications/updates. This discussion should continue in parallel.

**Comment by jjaswanson4** [comment]
Red Hat approves, with the following comments:

- We're wrapped around the axle around "gitops" vs. "api", and we shouldn't be. GitOps is a methodology, an API is a transport mechanism (in this case). GitOps is a set of practices that allow a code repository to define an actual deployment, requiring an orchestrator to be responsible for reconciling code vs. reality. An API, specifically a k8s-compliant API is the transport mechanism used to handle the reconciliation. It is our opinion that GitOps is a strong, valid way to drive and manage application deployments, but this functionality isn't required, and probably should be functionality added to an orchestrator.
- Deployment to non-k8s devices isn't clearly laid out here - what's the transport mechanism, etc? We need to consider defining a "margo agent", and having it be responsible for pulling deployment information from an orchestrator and running podman-compose/docker-compose, as a standard way to transport deployments to non-k8s devices.

**Comment by Silvanoc** [comment]
We could have an answer to this discussion that somehow is both GitOps and REST-API:
- [Flux has shown](https://sestegra.medium.com/gitops-with-flux-leveraging-oci-registry-as-a-state-store-ffe28dd64836) that you do not need Git to do GitOps on K8s.
  - GitOps is only a methodology, not bound to a specific tool (according the CNCF OpenGitOps project). I think this is what @jjaswanson4 is also mentioning.
- Doing GitOps with OCI registries as state stores ends up interacting with a REST-API.
  - No need for a somehow complex git client on the edge devices, a very simple REST-client suffice.
- Only half of the way to go:
  - Doing GitOps on K8s is already there, so no need to do anything.
  - Doing GitOps on docker-compose would be to be done.

We are already using the K8s custom resource definition format for the [desired-state-api](https://specification.margo.org/margo-api-reference/workload-api/desired-state-api/desired-state/), which is somehow a declarative way for operations. So getting the equivalent to Flux for docker-compose using OCI registries instead of Git should not be that difficult.

On the edge side we only need REST-clients talking to the [OCI registries REST-API](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#api) and YAML parsing.

It is important to note that this proposal from @phil-abb is not compatible with what is being proposed in [this page](https://github.com/margo/woa-interfaces-wg/blob/main/git-alternatives/proposals/desired_state_app_hash.md).

**Comment by jjaswanson4** [comment]
This piece is what we were discussing internally:
> Doing GitOps on docker-compose would need to be done

I think you're correct in calling out that it wouldn't be that difficult to actually implement, and we can discuss "what the orchestrator should do/prepare for the devices" and "what the devices should do themselves".

This does come with the caveat: we'd have two different "processes" for application deployments, one for k8s-compliant targets, and this new one for "non-k8s-compliant" targets, but honestly, I think that's fine. It also opens the door for other processes that could be device specific, if a vendor wants to contribute/implement their own delivery mechanisms to specific targets, such as embedded devices, that'd be fine as well, as long as they follow the standard.

Really, what I think I'm saying is: I'm don't necessarily care _how_ the deployment (or instructions) are delivered to the actual target, as long as we're all speaking the same language, so any orchestrator can be used with any end device, assuming both "speak the standard".

I don't think this is a blocker on Release 1, but we should look at @phil-abb 's proposal again, and chat through our options.

**Comment by pauldbrooks** [comment]
Josh, I like where you are going here (but defer to SMEs who know more).  The other question that this raises for me is the one of mandatory and optional behaviour.  Taking the vendor-specific workload that you described, to we expect that all WOS tools can deploy it, or would we consider it to be fair game for differentiation?  e.g. my k8s only WOS costs $x, but I have an option module that costs $y that supports non-k8s workloads.  A vendor of non-k8s workloads and targets will expect standardisation so many WOS tools can deploy, but may not be able to assume that every WOS tool can deploy

**Comment by phil-abb** [comment]
@pauldbrooks I think how WOS vendors choose to make money is outside Margo's scope. Right now we're saying the WOS has to comply with the Margo specification which includes the ability to deploy to devices running either Kubernetes or something like podman/docker to provide the interoperability Margo is trying to achieve. If a WOS vendor wishes to charge different fees to support each of them, then that would be up to them.

It might be worth discussing if we want to have different compliance tiers for WOS compliance. We've talked about this for device vendor compliance (e.g., optional device orchestration) but I don't know if it has been discussed for WOS vendors. 

I don't know if we'd want to break up WOS compliance to target the individual deployment profile we'll have though. Right now we're talking about helm, docker compose, and WASM but there will be more in the future. This seems to have the potential to become very confusing if we open the door for WOS vendors to pick and choose what deployment profiles they want to support.

**Comment by phil-abb** [comment]
@jjaswanson4 / @Silvanoc 

This is something to discuss on Tuesday's call but I don't see how basing this on an OCI Registry instead of a Git repository addresses the concerns raised by different companies. The only difference I see is that OCI has a specification and Git doesn't.

That still leaves the other concerns people have raised:
1. Having a different server you have to communicate with which means additional networking configuration on top of what is required to communicate with the WOS.  
2. Concerns over the number of calls the device has to make to get the desired state
3. Concerns over forcing a WOS vendor to use a specific way to maintain the desired state on the WOS side.

There are probably others that I'm missing as well.

From the Margo perspective, we only care about how (and what) the device and the WOS communicate, we don't care how either one implements any logic when it comes to the WOS maintaining and the device applying the desired state.

Also, the device needs to be able to get the desired state from the WOS "out of the box." Meaning, that the device vendor needs to be able to implement everything required to pull the desired state from any WOS vendor and include it when the device ships. We cannot require any new software to be deployed to the device once it is on the client site to enable the device to pull the desired state from the client's chosen WOS vendor.

**Comment by Silvanoc** [comment]
@phil-abb unfortunately I won't be able to join today.

In general you are right that this is probably a topic we should discuss in a call better than over GitHub issue comments. Giving that way others the opportunity to put questions.

**Comment by pauldbrooks** [comment]
> @pauldbrooks I think how WOS vendors choose to make money is outside Margo's scope. Right now we're saying the WOS has to comply with the Margo specification which includes the ability to deploy to devices running either Kubernetes or something like podman/docker to provide the interoperability Margo is trying to achieve. If a WOS vendor wishes to charge different fees to support each of them, then that would be up to them.
> 
> It might be worth discussing if we want to have different compliance tiers for WOS compliance. We've talked about this for device vendor compliance (e.g., optional device orchestration) but I don't know if it has been discussed for WOS vendors.
> 
> I don't know if we'd want to break up WOS compliance to target the individual deployment profile we'll have though. Right now we're talking about helm, docker compose, and WASM but there will be more in the future. This seems to have the potential to become very confusing if we open the door for WOS vendors to pick and choose what deployment profiles they want to support.

@phil-abb how WOS vendors make money is absolutely out of scope - that was not my intent.  My intent was to open a discussion about mandatory and optional standardised features.  Because if we force app/WOS/device vendors to support everything then we damage the ability to adopt which is absolutely our responsibility.

**Comment by ajcraig** [comment]
During our Workload Orchestration Agent ad-hoc technical discussion yesterday we discussed the current deployment patterns and an alternative approach was brought up via @jjaswanson4. 

Current approach
1. Deployment specification is created via the WOS service. This deployment specification includes:
    - Pointer to the helm chart location 
    - Configurable parameters and their instance specific answers 
2. Margo workload management agent then pulls the deployment specification from the wos.
3. Local helm client pulls/installs the charts on the cluster. Helm takes the packages chart, merges it with configuration values, and deploys the resulting Kubernetes resources to the cluster. 

Proposed / OpenGitops approach is a bit different. 
1. Helm chart is ingested by the OpenGitops tool, such as ArgoCD residing within the WOS.  
2. User assigns where the helm chart should be applied, typically a remote cluster that has been registered to the OpenGitops tool. 
3. ArgoCD renders the helm chart locally and merges it with configuration values. 
3. ArgoCD then sends the resulting Kubernetes resources to the cluster to be applied on the remote cluster.

Drawing snipped below: 
![image](https://github.com/user-attachments/assets/5435f6da-1980-46fc-a19b-f2036253edf5)



I wanted to bring this to everyone's attention as we close out the scope of Preview release 1. This would be a different approach to workload deployment, however I am not sure it completely breaks the pattern we have established. 

Options I see moving forward:
1. Allow WOS vendors to render the helm chart within their services and then utilize the Margo Management Interface to prepare only the rendered Kubernetes resources to the cluster via the Deployment specification
    - This method would put more responsibility on the WOS vendor and "lighten" the services required on the edge. Simply using Kubectl to apply the resulting rendered Kubernetes resources. 
2. Maintain our current approach with helm services residing on the edge where it pulls, renders, and applies all locally. 
3. Allow both options detailed above. This could allow differentiating features to be built within the WOS vendors' solutions. 

This topic will be discussed in our Thursday call on Dec 5th. 

Additional note: The ArgoCD tooling does allow for more than workloads to be managed via this state seeking mechanism. It allows for additional configurations and target states of other aspects of the cluster to be declared in the same manor, which ventures into our future scope of "Device orchestration".

**Comment by jjaswanson4** [comment]
Adding this to support ongoing discussions on point 1 above:

![Screenshot from 2024-12-05 09-29-01](https://github.com/user-attachments/assets/e4b96964-952c-4fb1-896e-dbca270d3496)

**Comment by phil-abb** [comment]
@ajcraig / @jjaswanson4 

I have several concerns about this new approach we need to work through:

- With our current approach, the WOS is intentionally abstracted away from the deployment type. This means the WOS doesn't need to know, or care, how an application gets installed/updated on the device, it only needs to tell the device what to deploy. Because of this, the current solution lets the WOS support any future deployment types with minimal, if any, changes required. With this new approach, we're forcing the WOS to implement adoption of these different deployment types to support them and WOS vendors may not see business value in taking the time to do that for all deployment types.

- This new proposal is a Kubernetes-specific solution. While some of the arguments make sense from a Kubernetes perspective it doesn't make as much sense for docker-compose, native WASM, and whatever future deployment types we support. This means we're inventing processes just so we can treat other deployment types similarly to helm deployments. It doesn't make sense to force docker-compose deployments to follow this same pattern, just to follow the pattern. At the same time, it doesn't make sense to have a completely different workflow just for Kubernetes. This is where I see the value in the original proposal, all deployment types are handled the same.

- We've discussed this a lot previously, we feel most end-users want to host the helm charts, docker-compose packages, and docker images locally for various reasons. There is very little chance end-users will want to expose this to a WOS outside the company's network. This means the end-user must upload the helm charts, docker-compose files, and files for whatever future deployment type we have for this new approach to work. This means these files are now outside the control of the end-user. On top of this, it means the WOS is now responsible for storing, and managing, all of these files for all their customers.

- For Kubernetes what is being suggested makes sense because the result is a big file with YAML. There are no other artifacts. With docker-compose, this often isn't the case. There are usually other supporting files packaged up along with the docker-compose.yml file which would now need to be transported from the WOS to the device. Since we don't know what future deployment types we will have there is the potential for this to apply to other deployment types as well.

- While it's not currently in the Margo specification, we have a use case where we want the device to be able to provide parameter values for things it doesn't make sense for an end-user to provide through the WOS UI. With this new approach, it means the device must send all of this information to the WOS but some of this information shouldn't leave the device.

- As mentioned in the call today, with the current approach end-users know the helm chart and docker-compose files being deployed are what they intended to deploy. If we change this to where the WOS is creating the final artifacts there is no good way of knowing the WOS hasn’t changed the artifacts in a way that wasn't intended. This doesn't have to be malicious, it could just be something the WOS may choose to do to support some differentiating feature they have.

- With the current approach, if there is a problem trying to install an application on a device you have two vendors involved. The device vendor and the app vendor. With this new approach, you have three vendors involved making support cases even more difficult. Trying to solve customer issues when two independent vendors are involved will be difficult, having three independent vendors involved will be nearly impossible.

- Several times we mentioned this new approach was better for small/constrained devices. While I agree with this sentiment, constrained devices are currently out of scope. The device we're talking about needs to be able to run all the other software components (OTEL collector, policy agent, etc.). I don't see the rending of these templates as that much of an overhead considering how infrequently it will happen, especially with smaller devices that can only run a few applications. When it comes time for supporting constrained devices something like this may make more sense but chances are these constrained devices will not even be running Kubernetes anyway.

- Helm has the [lookup function](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-lookup-function) and [chart hooks](https://helm.sh/docs/topics/charts_hooks/). Neither of these will work with the new approach since they require server-side rendering so we would have to say as part of Margo we do not support Helm charts that use these.

**Comment by jjaswanson4** [comment]
@ajcraig can correct me, but I actually think point 3 in his comment above is where we're currently sitting. This new approach doesn't go against anything that's already been defined, it simply shifts some responsibility around.

My main concerns around having devices do all the rendering are: 
- Limited compute resources available on the target devices, requiring additional compute for rendering
- Having devices in highly protected networks reaching out to many different places to deploy applications: X for the helm chart or docker-compose file, Y for the container images, Z for the margo.yaml file, etc, instead of centralizing as much of that as possible, and having more ability to limit "what can talk to what" at the network level
- The size/scale of the WOA, having to ship with a helm binary, something to edit/modify docker-compose files with information from margo.yaml files, an API client to talk to k8s APIs, another client to talk to the docker socket - and if we choose to add more deployment profiles, additional tools/libraries would need to be added, which could cause a lot of challenges when deployed to 1000s/10000s of devices with that much content

With that, specific responses:
> With our current approach, the WOS is intentionally abstracted away from the deployment type. This means the WOS doesn't need to know, or care, how an application gets installed/updated on the device, it only needs to tell the device what to deploy. Because of this, the current solution lets the WOS support any future deployment types with minimal, if any, changes required. With this new approach, we're forcing the WOS to implement adoption of these different deployment types to support them and WOS vendors may not see business value in taking the time to do that for all deployment types.

For the two defined deployment profiles, we're asking the WOS to do deployment rendering, because we know they involve templating to achieve the desired deployment.

Nothing prevents us from including a deployment type of "custom/none/etc", where the WOS doesn't render anything, and instead just passes everything "raw" to the WOA - which is, essentially, number 2 above.

I really don't like the idea of having to ship the helm binary, or another templating engine or code, plus supporting code, at the WOA level, to every device - I'd prefer the WOA simply download "what to do", and push it to the API or socket on the device.

This also allows vendors to implement specific deployment profiles for specific devices/etc in the future, where the WOS won't necessarily do any rendering, but will expect the WOA to know how to "talk" to specific devices that are not an API or socket.

> This new proposal is a Kubernetes-specific solution. While some of the arguments make sense from a Kubernetes perspective it doesn't make as much sense for docker-compose, native WASM, and whatever future deployment types we support. This means we're inventing processes just so we can treat other deployment types similarly to helm deployments. It doesn't make sense to force docker-compose deployments to follow this same pattern, just to follow the pattern. At the same time, it doesn't make sense to have a completely different workflow just for Kubernetes. This is where I see the value in the original proposal, all deployment types are handled the same.

Above, I have a flow outlined for docker-compose, as well as helm - both types are covered. However, there is already some gravity towards the "helm model", by having `parameters` and `configuration` in the margo.yaml application package definition. Those fields, in a way, act like a values.yaml file when using helm. With docker-compose, generally, the end user is asked to edit it according to their needs: adjusting env vars, or changing ports, etc, so there's already a need to do some templating against the application package definition to reach an application deployment - all that's happening here is shifting where that happens, up to the WOS, instead of relying on the WOA.

> For Kubernetes what is being suggested makes sense because the result is a big file with YAML. There are no other artifacts. With docker-compose, this often isn't the case. There are usually other supporting files packaged up along with the docker-compose.yml file which would now need to be transported from the WOS to the device. Since we don't know what future deployment types we will have there is the potential for this to apply to other deployment types as well.

What other files are you referring to? Startup files, binaries? Ideally, the only other "files" that need to be downloaded are built into container images, which the WOA isn't responsible for - the container engine (podman, docker) would be ultimately responsible for downloading those.

> While it's not currently in the Margo specification, we have a use case where we want the device to be able to provide parameter values for things it doesn't make sense for an end-user to provide through the WOS UI. With this new approach, it means the device must send all of this information to the WOS but some of this information shouldn't leave the device.

I'd like to learn more about this use case, but I believe we have this covered already under the device capabilities API. If there's specific information that needs to be used in rendering, the device can send it to the WOS, where it'll be available when rendering deployments.

We'd also be able to support "non-visible" data from devices as well - encrypted information that would never be visible to the end user - if there's sensitive information, although I think we'd need to come up with a specific example for this. I'm not sure what we'd consider highly sensitive data from the devices that Margo really cares about, since we're primarily interested in "what this device can do" more than anything.

> As mentioned in the call today, with the current approach end-users know the helm chart and docker-compose files being deployed are what they intended to deploy. If we change this to where the WOS is creating the final artifacts there is no good way of knowing the WOS hasn’t changed the artifacts in a way that wasn't intended. This doesn't have to be malicious, it could just be something the WOS may choose to do to support some differentiating feature they have.

Isn't this also true for the WOA today? Couldn't the WOA download a docker-compose file, change it, then deploy the wrong thing?

That scenario is, IMO, harder to identify overall, as the end user would need to be checking devices individually, at scale, over simply accessing the WOS and determining the exact definition that was pulled and applied by a WOA.

I'm all for having checks and balances to ensure what's actually desired is what's actually being deployed, either by checksums/hashes of the rendered deployments being compared, or introducing cryptographic protections, or any other mechanisms we choose.

> Several times we mentioned this new approach was better for small/constrained devices. While I agree with this sentiment, constrained devices are currently out of scope. The device we're talking about needs to be able to run all the other software components (OTEL collector, policy agent, etc.). I don't see the rending of these templates as that much of an overhead considering how infrequently it will happen, especially with smaller devices that can only run a few applications. When it comes time for supporting constrained devices something like this may make more sense but chances are these constrained devices will not even be running Kubernetes anyway.

I think we've settled on "embedded" or "non-generic compute devices" being out of scope, meaning purpose-built systems that can't run containers. Constrained devices, in this context, is simply "devices with limited compute resources", meaning smaller, lower powered x86 or ARM devices - think rPIs, single or dual core systems, whatever. Those are in scope, and will probably be used to run a single application.

Here's an example: a 2-core system, running two instances of a software-defined control runtime and control loop - each pinned to a core. For a device such as this, it'd be far less work, and far less disruptive to the running applications, to have the WOA be doing less work locally, as to avoid splitting the cores off from their main task, which is running the control runtime. This is compounded by requiring OTEL.

This is independent of kubernetes as well, same applies for docker-compose, or for other deployment profiles in the future.

> Helm has the lookup function and chart hooks. Neither of these will work with the new approach since they require server-side rendering so we would have to say as part of Margo we do not support Helm charts that use these.

1. I consider the information that would be "looked up" to be something we'd want the device to send back to the WOS - offloading the "heavy lifting" to the WOS, instead of the WOA/device. Above, I have boxes that outline "gather logs/information" - I'd honestly expect the WOA to be sending back "here's the resulting deployment after I threw it against the k8s API" - lookups could be performed against that, or going even further, the WOS could be "diffing" that against what it rendered, to ensure success/consistency.
2. Those aren't concepts in docker-compose, which is more bespoke in nature, so it'd be hard to maintain consistency between the two deployment profiles. Sure, docker-compose deployments don't really need to do lookups for services, since that's not a construct when running outside of k8s, but then what's being looked up in the k8s version of the app, and why?

**Comment by effndc** [comment]
The security concerns for `reaching out to many different places to deploy applications:  for the helm chart or docker-compose file, Y for the container images, Z for the margo.yaml file, etc` are not inherent in the current design, we just do not prevent the disaggregation.  The solution vendor can easily consolidate and host all of those assets to be stored in a single trusted location or use an OCI proxy (e.g. [helm-charts-oci-proxy](https://github.com/container-registry/helm-charts-oci-proxy) or [Harbor as a proxy](https://goharbor.io/docs/2.1.0/administration/configure-proxy-cache/)) if needed for their target use cases or customers.

Are we losing fundamental benefits of `helm` and ultimately requiring duplicating features of Helm in order to not have the packages on a Device?  WOA would have to recreate the equivalent of `helm test`, `helm upgrade`, `helm rollback`, after WOS translated them into some new formats for sending to WOA.

We would be duplicating Helm functionality to resolve dependencies, managing the complexity of chart hooks, complex aggregate charts, etc.  I am not aware of existing work that would perform this resolution and expansion of the Helm actions into a format that would fit into this architecture, as Helm itself was intended to be performed at the end point where applications are deployed.

**Comment by phil-abb** [comment]
@jjaswanson4 

> Having devices in highly protected networks reaching out to many different places to deploy applications: X for the helm chart or docker-compose file, Y for the container images, Z for the margo.yaml file, etc, instead of centralizing as much of that as possible, and having more ability to limit "what can talk to what" at the network level

This would be up to the end-user to locate these resources wherever they choose. Most likely for Kubernetes the helm chart and docker images will be located in the same OCI registry the end-user controls. The end-user still needs the OCI registry to host the docker images so it's nothing additional. We haven't discussed much about how the docker-compose files will be packaged and hosted yet. The Margo.yaml would still be hosted in the same place regardless of the direction we go with this. Based on the [current specification this comes from a git repository](https://specification.margo.org/app-interoperability/workload-orch-to-app-reg-interaction/). Or were you referring to the [ApplicationDeployment](https://specification.margo.org/margo-api-reference/workload-api/desired-state-api/desired-state/) (desired state) file?


> The size/scale of the WOA, having to ship with a helm binary, something to edit/modify docker-compose files with information from margo.yaml files, an API client to talk to k8s APIs, another client to talk to the docker socket - and if we choose to add more deployment profiles, additional tools/libraries would need to be added, which could cause a lot of challenges when deployed to 1000s/10000s of devices with that much content

> I think we've settled on "embedded" or "non-generic compute devices" being out of scope, meaning purpose-built systems that can't run containers. Constrained devices, in this context, is simply "devices with limited compute resources", meaning smaller, lower powered x86 or ARM devices - think rPIs, single or dual core systems, whatever. Those are in scope, and will probably be used to run a single application.

We need to ensure we are all aligned on this because this is not my understanding of how things are currently defined. Based on the [current specification](https://specification.margo.org/device-interoperability/device-requirements/) we have:

- Roles for `cluster leader` and `cluster worker` for multi-node clusters for helm deployments. Currently, we are only discussing the support for multiple node clusters from the same vendor. Multi-node cluster support is not in scope for the preview release under discussion. Discussions about supporting multi-node clusters composed of devices from multiple vendors have been discussed but is deferred for now.
- Role for `standalone cluster` for single node Kubernetes for helm deployments.
- Role for `standalone device` for a device that handles docker-compose deployments.

Each of these has a list of software that needs to be installed on the device for it to be considered Margo-compliant.

Based on previous discussion we have said a device fills only one role so a device only supports one deployment type.


`Micro` devices are devices that cannot fill one of the roles above. We have recognized the need to support these devices but we haven't spent much time discussing these yet.

I know there is an ongoing taxonomy discussion in the steering committee to improve on how these devices are classified but this has not been presented to the technical working group yet.

> What other files are you referring to? Startup files, binaries? Ideally, the only other "files" that need to be downloaded are built into container images, which the WOA isn't responsible for - the container engine (podman, docker) would be ultimately responsible for downloading those.

Configuration files are the biggest example. Kubernetes has configmap manifest so you can include them in the helm chart. Docker does not, so you must include your config files in the package so they can be mounted when starting the container. There could be any number of other files that could be included, for different reasons, that are not included in the image because they don't belong in the image. This is not an uncommon practice for docker-compose deployments.

> Nothing prevents us from including a deployment type of "custom/none/etc", where the WOS doesn't render anything, and instead just passes everything "raw" to the WOA - which is, essentially, number 2 above.

Technically this is true, but previously we have tried to avoid having multiple ways of doing things because either the WOS, the device, or both must implement all options which adds complexity. That's not to say we cannot have multiple ways of doing things, we just need to agree the overhead is worth it.
